rfcs/0049-flakes.md

```diff
@@ -0,0 +1,381 @@
+
+            cp dwarffs $out/bin/
+            ln -s dwarffs $out/bin/mount.fuse.dwarffs
+
+            cp ${./run-dwarffs.mount} $out/lib/systemd/system/run-dwarffs.mount
+            cp ${./run-dwarffs.automount} $out/lib/systemd/system/run-dwarffs.automount
+          '';
+      };
+
+    nixosModules.dwarffs = ...;
+
+    defaultPackage = packages.dwarffs;
+
+    checks.build = packages.dwarffs;
+  };
+}
+```
+
+A flake has the following attributes:
+
+* `edition`: A number that specifies the version of the flake
+  syntax/semantics to be used. This allows the interpretation of
+  flakes to change in the future. It also enables some evolution of
+  the Nix language; for example, the Nix files in the flake could be
+  parsed using a syntax determined by the edition. The only currently
+  allowed value is `201911`. Nix rejects flakes with an unsupported
+  edition.
+
+* `description`: A short description of the flake.
+
+* `inputs`: An attrset specifying the dependencies of the flake
+  (described below).
+
+* `outputs`: A function that, given an attribute set containing the
+  outputs of each of the input flakes keyed by their identifier,
+  yields the Nix values provided by this flake. Thus, in the example
+  above, `inputs.nixpkgs` contains the result of the call to the
+  `outputs` function of the `nixpkgs` flake, and
+  `inputs.nixpkgs.packages.fuse` refers to the `packages.fuse` output
+  attribute of `nixpkgs`.
+
+  In addition to the outputs of each input, each input in `inputs`
+  also contains some metadata about the inputs. These are:
+
+  * `outPath`: The path in the Nix store of the flake's source
+    tree. This means that you could import Nixpkgs in a more
+    legacy-ish way by writing
+
+        with import inputs.nixpkgs { system = "x86_64-linux"; };
+
+    since `nixpkgs` still contains a `/default.nix`. In this case we
+    bypass its outputs entirely and only use the flake mechanism to
+    get its source tree.
+
+  * `rev`: The commit hash of the flake's Git repository.
+
+  * `revCount`: The number of ancestors of the revision `rev`. This is
+    not available for `github` repositories (see below), since they're
+    fetched as tarballs rather than as Git repositories.
+
+  * `lastModified`: The commit time of the revision `rev`, in the
+    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
+    this is available for both Git and GitHub repositories, so it's
+    useful for generating (hopefully) monotonically increasing version
+    strings.
+
+  The value returned by the `outputs` function must be an attribute
+  set. The attributes can have arbitrary values; however, some tools
+  may require specific attributes to have a specific value (e.g. the
+  `nix` command may expect the value of `packages` to be an attribute
+  set of derivations).
+
+## Flake inputs
+
+The attribute `inputs` specifies the dependencies of a flake. These
+specify the location of the dependency, or a symbolic flake identifier
```

Comment:
Will it be possible to force-override an input location (with the user being responsible for validity of such subsitution)?

Reply:
Yes.

Reply:
Maybe mention it somewhere?

Reply:
Well, I removed all user interface aspects from the RFC :-) But it's something like `--override-flake OLD NEW`.

Reply:
Ah, I underestimated the completeness of removal of UI mentions. I thought it is about «how something is done», but some of the «what should be possible to do» is still specified.